src/crimson/osd/shard_services.cc

```diff
@@ -101,6 +101,22 @@ seastar::future<> ShardServices::send_to_osd(
   }
 }
 
+seastar::future<> ShardServices::send_to_osd(
+  int peer, MessageURef m, epoch_t from_epoch) {
+  if (osdmap->is_down(peer)) {
+    logger().info("{}: osd.{} is_down", __func__, peer);
+    return seastar::now();
+  } else if (osdmap->get_info(peer).up_from > from_epoch) {
+    logger().info("{}: osd.{} {} > {}", __func__, peer,
+		    osdmap->get_info(peer).up_from, from_epoch);
+    return seastar::now();
+  } else {
+    auto conn = cluster_msgr.connect(
+        osdmap->get_cluster_addrs(peer).front(), CEPH_ENTITY_TYPE_OSD);
+    return conn->send(std::move(m));
```

Comment:
IIUC the main difference between these two variants is just that. Maybe it worth to unify the methods by e.g. introducing a third, template variant (it would abstract the message ptr type and always do `std::move`) called from them.